t/integration/test_canvas.py

```diff
@@ -859,6 +859,154 @@ def before_task_publish_handler(sender=None, body=None, exchange=None, routing_k
         redis_connection = get_redis_connection()
         redis_connection.delete(redis_key)
 
+    def test_chaining_upgraded_chords_pure_groups(self, manager, subtests):
+        """ This test is built to reproduce the github issue https://github.com/celery/celery/issues/5958
```

Comment:
the integration tests looks good so far.

Reply:
Thanks!
These tests cover:
1. Fix in chord.__or__.
2. Fix in chain.__or__.
3. Make sure the workaround from the last 2 years is not broken from this change.